src/render/font.rs

```diff
@@ -0,0 +1,59 @@
+// Font data translated from X11 font, tool forthcoming.
```

Comment:
Another area that is hard to understand is the representation of the font data. I understand that the pixel data is 1bpp. So is font stride = 92 bits? and font height 15 bits? That would make each glyph 6.13 bits tall, which doesn't make sense.

Maybe once I am able to see the tool it will make a bit more sense, but right now this is a bit of a black box to me.

Reply:
Yeah, I'm going to redo it a bit, largely so that space doesn't need an op (and also to enable a bit more kerning), which is one reason it's not fully baked. But I'll give a little explanation now.

The stride is 92 bytes, so this encodes an image 92 * 8 pixels wide by 15 pixels tall. Glyphs are variable width but all the same height. The glyphs are packed so that they don't cross 32-bit boundaries, so that to render a glyph only a single 32 bit read is needed (per scan line). The current code has a limitation that glyphs can't be wider than 32 bits, but that could be generalized without too much trouble.

To render a glyph, you read a 32 bit value, mask it (deleting other glyphs packed into the same word), and shift it into place. Because of the particulars of ARM, that's most efficiently done as a sequence of shift-right, shift-left, shift-right. The multiple shift amounts are packed up to 4 per 32-bit word. This all results in super-efficient asm, which would be hard to represent in Rust without loss of efficiency, as the semantics for shift values >= 32 don't quite match up.

Reply:
Having this explanation makes it a lot clearer. That is a very smart way of storing the fonts. It fits closely with what I know from other tools' font rendering (mainly egui) where they have a big texture with all of the glyphs, and way to extract each glyph. The method here just has the extra complexity of the glyphs being bit-packed.